open-api/rest-catalog-open-api.yaml

```diff
@@ -3642,6 +3781,173 @@ components:
+        - $ref: '#/components/schemas/IncrementalSnapshotScanContext'
+
+    BaseTableScanContext:
+      discriminator:
+        propertyName: table-scan-type
+        mapping:
+          snapshot-scan: '#/components/schemas/SnapshotScanContext'
+          incremental-snapshot-scan: '#/components/schemas/IncrementalSnapshotScanContext'
+      type: object
+      required:
+        - table-scan-type
+      properties:
+        table-scan-type:
+          type: string
+
+    SnapshotScanContext:
+      description: context for scanning data in a specific snapshot
+      type: object
+      allOf:
+        - $ref: '#/components/schemas/BaseTableScanContext'
+      required:
+        - table-scan-type
+      properties:
+        table-scan-type:
+          type: string
+          enum: ["snapshot-scan"]
+        select:
+          $ref: '#/components/schemas/SelectedFieldNames'
+        filter:
+          $ref: '#/components/schemas/Filter'
+        case-sensitive:
+          description: If field selection and filtering should be case sensitive
+          type: boolean
+          default: true
+        snapshot-id:
+          description:
+            The ID of the snapshot to use for the table scan.
+            If not specified, the snapshot at the main branch head will be used.
+          type: integer
+          format: int64
+        use-snapshot-schema:
+          description:
+            If the schema of the specific snapshot should be used instead of the table schema.
+          type: boolean
+          default: false
+
+    IncrementalSnapshotScanContext:
+      description:
+        Context for scanning data appended in a range of snapshots.
+        The scan always follows the schema of the snapshot at the main branch head.
+      type: object
+      allOf:
+        - $ref: '#/components/schemas/BaseTableScanContext'
+      required:
+        - table-scan-type
+        - start-snapshot-id
+      properties:
+        table-scan-type:
+          type: string
+          enum: ["incremental-snapshot-scan"]
+        select:
+          $ref: '#/components/schemas/SelectedFieldNames'
+        filter:
+          $ref: '#/components/schemas/Filter'
+        case-sensitive:
+          description: If field selection and filtering should be case sensitive
+          type: boolean
+          default: true
+        start-snapshot-id:
+          description: The ID of the starting snapshot of the incremental scan
+          type: integer
+          format: int64
+        inclusive-start:
+          description: If the data appended in the start snapshot should be included in the scan
+          type: boolean
+          default: false
+        end-snapshot-id:
+          description:
+            The ID of the inclusive ending snapshot of the incremental scan.
+            If not specified, the snapshot at the main branch head will be used as the end snapshot.
+          type: integer
+          format: int64
+
+    FieldName:
```

Comment:
I think this is what a previous comment mean about specifying the field name requirements?

Reply:
Are you talking about the comment from @rdblue regarding how to handle nested fields? I think it may have been his original coment https://github.com/apache/iceberg/pull/9695#discussion_r1494914340

Reply:
yes

Reply:
Hi @jackye1995 - first off, thank you very much for the work you are doing on this item!

I'm personally incredibly excited at the performance optimization this feature will enable Rest Catalogs to bring to data ecosystems everywhere.

I do have a question on the decision to use flat '.' delimited strings to represent FieldName here instead of an array.

While working on PRs on PyIceberg, we discovered that flattening the names has a potential to cause namespace conflicts in the field names. For example, this will cause a field named `a.b` to collide with the child field `b` of field `a`. This can be represented by an Iceberg Schema like:

```
Schema(
    NestedField("a.b", StringType(), id=1),
    NestedField("a", StructType(
        [NestedField("b", StringType(), id=3)]
    ), id=2),
)
```

Flattening the name actually contrasts our approach documented in `Column Projection` section of the docs, where we note that a name may contain '.' but that this refers to a literal name, and not a nested field: https://iceberg.apache.org/spec/#column-projection
 
Open API Schema for Namespace, which also represents a nested structure:
https://github.com/apache/iceberg/blob/4a0ae22199375e34f9033bed9781da3dc90d53c6/open-api/rest-catalog-open-api.yaml#L1625-L1630

Although it goes against our existing approach, I can see how flattening the names would be appealing given that maintaining the logic to create a nested structure and implement lookups in all of our libraries may be a more challenging task. But given that we already have documentation that says otherwise, I think if we decide to use a flat representation here, we should be intentional about it, and update our documentations to note that using '.' character in column names will result in errors with our field name projection

Reply:
Yes that's a good point, thanks for bringing it up! This logic was derived from https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/types/IndexByName.java, which does have the issue about escaping the `.` character because the quoting function is the identity function by default. I am wondering if we can develop a quoting function that works for REST use case. Spark uses a quoting function to surround names with backquote, so it can distinguish `a.b` from `` `a.b` ``, but that means column names with backquote would have problems. Maybe using the same escape character `%1F` as namespace would work? Although it's a bit ugly...

Reply:
I think the reason Namespace has a principle on flat representation is because there's a GET endpoint that requires us to use a flat representation. 

It looks like we only have POST endpoints for PreplanTable and PlanTable, so I'm wondering if we need to worry about the flat representation here.

Could we just choose to represent the `SelectedFieldNames` as an array of `FieldName` - which in turn can just be an `array of strings`?.

Reply:
@syun64, @jackye1995, @amogh-jahagirdar, flattening with `.` is not a problem for Iceberg tables because ambiguous names like `a.b` in the example type that Sung posted are not allowed.

This is how `Schema` has worked from very early on ([tests added in 2019](https://github.com/apache/iceberg/commit/8b41ee5e34b5d79e7a04b3c810d7073cfed01a3a)) and why `findField` has always used the name to ID index.

The rationale for this is that there are two ways to handle conflicts like this. First, you could use escaping and have extra complexity when parsing structures. Or second, you can disallow ambiguous names so that you always have a clear mapping from the flattened string. Iceberg opted for the second option for column names so that there is no need for escaping.

This has really helped simplify a lot of APIs! For instance, `Schema.select(String...)` can select multiple columns easily rather than needing to have a complicated signature like `Schema.select(List<String>...)`.

A couple more things to clarify:

> Flattening the name actually contrasts our approach documented in Column Projection section of the docs, where we note that a name may contain '.' but that this refers to a literal name, and not a nested field

This actually does not conflict. This section is documenting what the field name in a column mapping means, and this is saying that the name string identifies an immediate child and not a further nested child. This is because we do not index sub-sections of these tree structures. Iceberg only indexes fields by name in `Schema` and not in each `StructType`, for example.

> I think the reason Namespace has a principle on flat representation is because there's a GET endpoint that requires us to use a flat representation.

Actually, I originally advocated for the same approach for namespaces in URLs. I think we should have used `.` as the delimiter because it is easier to work with. The trade-off is that REST catalogs would not be able to support ambiguous names, like `["a", "b"]` in the same catalog as `["a.b"]`.

That's not what the community decided to go forward with, so we have escape characters when sending namespaces in URLs. But also note that namespaces in the REST spec use a JSON array of strings whenever that is possible.